src/Language/Haskell/LSP/Core.hs

```diff
@@ -674,14 +691,19 @@ initializeRequestHandler' onStartup mHandler tvarCtx = Handler $ \req rspH@(Clie
                               Cancellable -> True
                               NotCancellable -> False
 
-          rId <- freshLspId tvarCtx
-
           -- Create progress token
-          liftIO $ sf $ J.fromServerReq $ J.RequestMessage
-            "2.0" (J.IdInt rId) J.SWindowWorkDoneProgressCreate (J.WorkDoneProgressCreateParams progId)
+          -- FIXME  : This needs to wait until the request returns before
+          -- continuing!!!
+          _ <- mkSendReqFunc tvarCtx J.SWindowWorkDoneProgressCreate
+                (J.WorkDoneProgressCreateParams progId) $ \_ res -> do
+                  case res of
+                    -- An error ocurred when the client was setting it up
+                    -- No need to do anything then, as per the spec
+                    Left _err -> pure ()
+                    Right () -> pure ()
 
           -- Send initial notification
-          liftIO $ sf $ J.fromServerNot $ J.NotificationMessage "2.0" J.SProgress $
+          mkSendNotFunc tvarCtx J.SProgress $
```

Comment:
Couldn't these notifications be moved inside the `Right () -> ` case branch above to fix the issue?

Reply:
Unfortunately it needs to return the result of running the user's callback, which would have to happen inside the response handler. I think it might be possible with the use of some async